src/audio/codec_adapter/codec/dcblock.c

```diff
@@ -0,0 +1,383 @@
+ * H(z) = (1 - z^-1)/(1-Rz^-1).
+ * Setting R to 1 makes the filter act as a passthrough component.
+ */
+static void set_passthrough(struct dcblock_codec_data *cd)
+{
+	int i;
+
+	for (i = 0; i < PLATFORM_MAX_CHANNELS; i++)
+		cd->R_coeffs[i] = ONE_Q2_30;
+}
+
+static int apply_config(struct comp_dev *dev, enum codec_cfg_type type)
+{
+	int ret;
+	int size;
+	struct codec_config *cfg;
+	void *data;
+	struct codec_param *param;
+	struct codec_data *codec = comp_get_codec(dev);
+	struct dcblock_codec_data *cd = codec->private;
+	size_t req_size;
+
+	comp_dbg(dev, "apply_config() start");
+
+	cfg = (type == CODEC_CFG_SETUP) ? &codec->s_cfg :
+					  &codec->r_cfg;
+	data = cfg->data;
+	size = cfg->size;
+
+	if (!cfg->avail || !size) {
+		comp_err(dev, "apply_config() error: no config available, requested conf. type %d",
+			 type);
+		ret = -EIO;
+		goto ret;
+	}
+
+	/* Read parameters stored in `data` - it may keep plenty of
+	 * parameters. The `size` variable is equal to param->size * count,
+	 * where count is number of parameters stored in `data`.
+	 */
+	while (size > 0) {
+		param = data;
+		comp_dbg(dev, "apply_config() applying param %d size %d value[0] %d",
+			 param->id, param->size, param->data[0]);
+		/* Set read parameter */
+		if (param->id == DCBLOCK_CONFIG_R_COEFFS) {
+			req_size = sizeof(cd->R_coeffs);
+			if (param->size < req_size + sizeof(param->id) + sizeof(param->size)) {
+				comp_err(dev, "apply_config() error: parameter size %d not enough, required %d",
+					 param->size, req_size);
+				ret = -EINVAL;
+				goto ret;
+			}
+			ret = memcpy_s(cd->R_coeffs, req_size, param->data, req_size);
+			if (ret != 0)
+				goto ret;
+		} else {
+			comp_err(dev, "apply_config() error: parameter id %d not exists", param->id);
+			ret = -EINVAL;
+			goto ret;
+		}
+		/* Obtain next parameter, it starts right after the preceding one */
+		data = (char *)data + param->size;
+		size -= param->size;
+	}
+
+	comp_dbg(dev, "apply_config() done");
+ret:
+	return ret;
+}
+
+int dcblock_codec_init(struct comp_dev *dev)
+{
+	struct codec_data *codec = comp_get_codec(dev);
+	struct dcblock_codec_data *cd = NULL;
+
+	comp_info(dev, "dcblock_codec_init()");
+
+	cd = codec_allocate_memory(dev, sizeof(struct dcblock_codec_data), 0);
+	if (!cd) {
+		comp_err(dev, "dcblock_codec_init(): failed to allocate memory for dcblock codec data");
+		return -ENOMEM;
+	}
+
+	codec->private = cd;
+
+	memset(cd->state, 0, sizeof(cd->state));
+	set_passthrough(cd);
+	cd->dcblock_func = NULL;
+
+	comp_dbg(dev, "dcblock_codec_init() done");
+
+	return 0;
+}
+
+int dcblock_codec_prepare(struct comp_dev *dev)
+{
+	int ret = 0;
+	struct codec_data *codec = comp_get_codec(dev);
+	struct dcblock_codec_data *cd = codec->private;
+	struct comp_buffer *sourceb;
+	enum sof_ipc_frame source_format;
+
+	comp_info(dev, "dcblock_codec_prepare()");
+
+	if (codec->state == CODEC_PREPARED)
+		goto done;
+
+	if (!codec->s_cfg.avail && !codec->s_cfg.size) {
```

Comment:
just curious why we are doing this here? I would think we should see if we have a config set to our local data otherwise fail? Should we really go out searching for the setup config since it will always call apply config?

Reply:
As what I've understood s_cfg (Setup Config) is the config which should set before CA gets PREPARED. It loads the default on codec_adapter_new() and may be updated on codec_adapter_set_params(). The codec then should be set from s_cfg on codec_prepare()

Reply:
Yea, im taking this up at the TSC, this API irks me, it feels like the adapter is trying to parent the codec rather than provide a thin connection path.